File: c++ version/1-100/52-n-queens-II.cpp
int totalNQueens(int n) {
    int count = 0;
    if (n == 0) 
        return count;
    vector<bool> column(n, false), ldiag(2*n-1, false), rdiag(2*n-1, false);
    backtracking(column, ldiag, rdiag, count, 0, n);
    return count;
}

void backtracking(vector<bool> &column, vector<bool> &ldiag,
    vector<bool> &rdiag, int &count, int row, int n) {
    if (row == n) {
        ++count;
        return;
    }
    for (int col = 0; col < n; ++col) {
        if (column[col] || ldiag[n-row-1+col] || rdiag[col+row])
            continue;
        column[col] = ldiag[n-row-1+col] = rdiag[col+row] = true;
        backtracking(column, ldiag, rdiag, count, row + 1, n);
        column[col] = ldiag[n-row-1+col] = rdiag[col+row] = false;
    }
}

File: python version/52-n-queens-II.py
class Solution:
    def totalNQueens(self, n: int) -> int:
        cols = [False] * n
        l_diags = [False] * (2*n-1)
        r_diags = [False] * (2*n-1)
        board = [["."]*n for _ in range(n)]
        self.count = 0
        def dfs(row):
            if row == n:
                self.count += 1
                return
            for col in range(n):
                if cols[col] or l_diags[row+col] or r_diags[n-1-col+row]: continue
                cols[col] = l_diags[row+col] = r_diags[n-1-col+row] = True
                board[row][col] = "Q"
                dfs(row+1)
                board[row][col] = "."
                cols[col] = l_diags[row+col] = r_diags[n-1-col+row] = False
        dfs(0)
        return self.count